infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/PreviewUrlCommandProvisioner.java

```diff
@@ -0,0 +1,115 @@
+import io.fabric8.kubernetes.api.model.extensions.IngressBackend;
+import io.fabric8.kubernetes.api.model.extensions.IngressRule;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import javax.inject.Singleton;
+import org.eclipse.che.api.workspace.server.model.impl.CommandImpl;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.KubernetesNamespace;
+import org.eclipse.che.workspace.infrastructure.kubernetes.util.Services;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Updates commands with proper Preview URL attribute.
+ *
+ * <p>Goes through all {@link CommandImpl}s in given {@link KubernetesEnvironment} and for ones that
+ * have defined Preview URL tries to find matching {@link Service} and {@link Ingress} from given
+ * {@link KubernetesNamespace}. When found, it composes full URL and set it as attribute of a
+ * command with key {@link
+ * org.eclipse.che.api.core.model.workspace.config.Command#PREVIEW_URL_ATTRIBUTE}.
+ *
+ * @param <E> type of the environment
+ */
+@Singleton
+public class PreviewUrlCommandProvisioner<E extends KubernetesEnvironment> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(PreviewUrlCommandProvisioner.class);
+
+  public void provision(E env, KubernetesNamespace namespace) throws InfrastructureException {
+    injectsPreviewUrlToCommands(env, namespace);
+  }
+
+  /**
+   * Go through all commands, find matching service and exposed host. Then construct full preview
+   * url from this data and set it as Command's parameter under `previewUrl` key.
+   *
+   * @param env environment to get commands
+   * @param namespace current kubernetes namespace where we're looking for services and ingresses
+   */
+  private void injectsPreviewUrlToCommands(E env, KubernetesNamespace namespace)
+      throws InfrastructureException {
+    if (env.getCommands() == null) {
+      return;
+    }
+    for (CommandImpl command :
+        env.getCommands()
+            .stream()
+            .filter(c -> c.getPreviewUrl() != null)
+            .collect(Collectors.toList())) {
+      Optional<Service> foundService =
+          Services.findServiceWithPort(
+              namespace.services().get(), command.getPreviewUrl().getPort());
+      if (foundService.isPresent()) {
+        Optional<String> foundHost =
+            findHostForServicePort(
+                namespace, foundService.get(), command.getPreviewUrl().getPort());
+        if (foundHost.isPresent()) {
+          command.getAttributes().put(PREVIEW_URL_ATTRIBUTE, foundHost.get());
+        } else {
+          LOG.warn(
+              "unable to find ingress for service [{}] and port [{}]",
+              foundService.get(),
+              command.getPreviewUrl().getPort());
+        }
+      } else {
+        LOG.warn(
```

Comment:
Personally I prefer to reduce nesting levels, so
Please consider the following change
```java
 for (CommandImpl command :
        env.getCommands()
            .stream()
            .filter(c -> c.getPreviewUrl() != null)
            .collect(Collectors.toList())) {
      Optional<Service> foundService =
          Services.findServiceWithPort(
              namespace.services().get(), command.getPreviewUrl().getPort());

      if (!foundService.isPresent()) {
        LOG.warn(
            "unable to find service for port [{}] for command [{}]",
            command.getPreviewUrl().getPort(),
            command.getName());
        return;
      }

      Optional<String> foundHost =
          findHostForServicePort(
              namespace, foundService.get(), command.getPreviewUrl().getPort());
      if (foundHost.isPresent()) {
        command.getAttributes().put(PREVIEW_URL_ATTRIBUTE, foundHost.get())
      } else {
        LOG.warn(
            "unable to find ingress for service [{}] and port [{}]",
            foundService.get(),
            command.getPreviewUrl().getPort());
      }
    }
```
but it's question of taste. Up to you change it or not.

Reply:
`return;` in first `if` should be `continue;` right? we're in the loop. That's what I wanted to avoid here, cutting the loop in the middle. I think 2 nested `if`s are ok here and are easier to follow. I usually prefer single return from the method or block. But I don't have strong preference here. If anyone else comment on this, I'll rewrite it, if not I'll keep it as is :))

Reply:
> return; in first if should be continue; right?

right!

Reply:
I would vote 55 to 45 for @sleshchenko variant. @metlos @mshaposhnik  wdyt?

Reply:
done